Initial implementation of paint effects (#880)
flame-engine/flame: doc/effects.md

```dart
final combination = CombinedEffect(
    effects: [move, scale, rotate],
    isInfinite: true,
    isAlternating: true);
myComponent.addEffect(combination);
```

An example of how to use the `CombinedEffect` can be found
[here](https://github.com/flame-engine/flame/tree/main/examples/lib/stories/effects/combined_effect.dart).

> ## Common for paint effects

Reviewer: I think instead of having `## Common for MoveEffect, ScaleEffect and RotateEffect (SimplePositionComponentEffects)` and this, we could have second-level `##` with the type of effect, eg `## Position Effects`, `## Paint Effects`, followed by a brief desc and second level `###` chapters for each type (here and above). Wdyt?

Author: Yeah, I think it could look better indeed, but I just followed the current structure. Changing to your proposed structure would make me need to change more things on the doc unrelated to this PR, so I think we could do it, but in a another PR. wdyt?

Reviewer: Sure, it can be a followup

> Flame provides an useful mixin that adds paint variables to your components, that is used by some
> of Flame's own components like `SpriteComponent`, `SpriteAnimationComponent`, but you can also use
> it on you any custom component that your game has, and any component that uses that mixin can have
> paint affects applied to it.

Reviewer: ```suggestion
Flame provides an useful mixin that adds paint variables to your components. This mixin is used by some
of Flame's own components like `SpriteComponent`, `SpriteAnimationComponent`, but you can also use
it on you any of your custom components. Any component that uses this mixin can have
paint effects applied to it.
```

> By default all effects will animate the main paint of the component, that can be changed by passing
> a `paintId` to the effect contructor. This can be useful for when you have a component that have
> multiple paints, for example a component which has a foreground and background layer.

Reviewer: ```suggestion
By default, all effects will animate the main paint of the component. That can be changed by passing
a `paintId` [note: or key?] to the effect contructor. This can be useful for when you have a component that have
multiple paints; for example, a component which has a foreground and background layer.
```

I think before this paragraph explaining how effects behaves wrt to multiple paints, we should have a paragraph dedicated to explaining how the `HasPaint` mixing allows for one main paint and other, optional sub-paints that you can key by providing a generic type parameter (because that is not super easy to grasp so it's good to be precise before even talking about effects).

> Check below for a list of the available paint effects

Reviewer: ```suggestion
Below is a list of the available paint effects.
```

## OpacityEffect

> This effect allows you animate the opacity of your paint, it receives a double for the opacity,
> which must be between 0 and 1 (including) and a duration in seconds, represented by a double as
> well.

Reviewer: ```suggestion
This effect allows you animate the opacity of your paint. It receives a double for the opacity,
which must be between 0 and 1 (including), and a duration in seconds, represented by a double as
well.
```

Usage example:

```dart
myComponent.addEffect(
  OpacityEffect(
    opacity: 0,
    duration: 0.5,
  ),
);
```

An example of how to use the `OpacityEffect` can be found
[here](https://github.com/flame-engine/flame/tree/main/examples/lib/stories/effects/opacity_effect.dart).

## ColorEffect

> This effect will change the base color of the paint, making things rendered with it, looks like
> it was tinted with the passed color.

Reviewer: ```suggestion
This effect will change the base color of the paint, causing the rendered component to be tinted by the provided color.
```

> Usage Example:

Reviewer: ```suggestion
Usage example:
```

consistency is key :p

```dart
myComponent.addEffect(
  ColorEffect(
    color: const Color(0xFF00FF00),
    duration: 0.5,
  ),
);
```

> An example of how to use the `ColorEffect` can be found

Reviewer: ```suggestion
A more in-depth example can be found
```
